gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsTracer.java

```diff
@@ -0,0 +1,164 @@
+    this.attributes.put("method_name", spanName.toString());
+    this.metricsRecorder = metricsRecorder;
+  }
+
+  @Override
+  public Scope inScope() {
+    return () -> {};
+  }
+
+  @Override
+  public void operationSucceeded() {}
+
+  @Override
+  public void operationSucceeded(Object response) {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.OK.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationCancelled() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.CANCELLED.toString());
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    attributes.put(STATUS_ATTRIBUTE, extractStatus(error));
+    metricsRecorder.recordOperationLatency(
+        operationTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordOperationCount(1, attributes);
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    //Ask Blakein how this works in comparison to baseApiTracer implementation 
+  }
+
+  @Override
+  public void attemptStarted(Object request, int attemptNumber) {
+    attemptTimer = Stopwatch.createStarted();
+  }
+
+  @Override
+  public void attemptSucceeded() {}
+
+  @Override
+  public void attemptSucceeded(Object response) {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.OK.toString());
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptCancelled() {
+    attributes.put(STATUS_ATTRIBUTE, StatusCode.Code.CANCELLED.toString());
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptFailed(Throwable error, Duration delay) {
+    attributes.put(STATUS_ATTRIBUTE, extractStatus(error));
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptFailedRetriesExhausted(Throwable error) {
+    // this is called in BasicRetryingFuture.java
+    // imo, ideally, I think this should put the status_attribute in the map, and attempt_count
+    // and attempt_latency should be calculated.
+
+    attributes.put(STATUS_ATTRIBUTE, extractStatus(error));
+    metricsRecorder.recordAttemptLatency(attemptTimer.elapsed(TimeUnit.MILLISECONDS), attributes);
+    metricsRecorder.recordAttemptCount(1, attributes);
+  }
+
+  @Override
+  public void attemptPermanentFailure(Throwable error) {
```

Comment:
Do we know if `operationFailed` would be called after `attemptPermanentFailure` or `attemptFailedRetriesExhausted`? 
Similarly, we need to make sure `operationCancelled` is called after `attemptCancelled`. If not, we need to record operation latency and operation count in these methods.

Reply:
Confirmed operationCancelled is called after attemptCancelled. Now testing the last two scenarios.